ASoC: SOF: Add support for on-demand DSP boot for LNL+

ujfalusi opened the pull request:
On system suspend / resume we always power up the DSP and boot the
firmware, which is not strictly needed as right after the firmware booted
up we power the DSP down again on suspend and we also power it down after
resume after some inactivity.

Out of caution, add a new platform descriptor flag to enable on-demand
DSP boot since this might not work without changes to platform code on
certain platforms.

With the on-demand dsp boot enabled we will not boot the DSP and firmware
up on system or rpm resume, just enable audio subsystem since audio IPs,
like HDA and SoundWire might be needed (codecs suspend/resume operation).
Only boot up the DSP during the first hw_params() call when the DSP is
really going to be needed.

In this way we can handle the audio related use cases:
normal audio use (rpm suspend/resume)
system suspend/resume without active audio
system suspend/resume with active audio
system suspend/resume without active audio, and audio start before the rpm
suspend timeout

Suspend resume is tested on LNL laptop with `pm-graph/sleepgraph.py -rtcwake 15 -m freeze`
Baseline (w/o active audio):
sof-audio-pci-intel-lnl @ 0000:00:1f.3 {sof-audio-pci-intel-lnl} (**Total Suspend: 106.434 ms Total Resume: 157.955 ms**)

With this PR, without active audio:
sof-audio-pci-intel-lnl @ 0000:00:1f.3 {sof-audio-pci-intel-lnl} (**Total Suspend: 18.351 ms Total Resume: 31.468 ms**)

With this PR, with active audio:
sof-audio-pci-intel-lnl @ 0000:00:1f.3 {sof-audio-pci-intel-lnl} (**Total Suspend: 116.535 ms Total Resume: 32.763 ms**)

ujfalusi committed:
ASoC: SOF: Add support for on-demand DSP boot

On system suspend / resume we always power up the DSP and boot the
firmware, which is not strictly needed as right after the firmware booted
up we power the DSP down again on suspend and we also power it down after
resume after some inactivity.

Out of caution, add a new platform descriptor flag to enable on-demand
DSP boot since this might not work without changes to platform code on
certain platforms.

With the on-demand dsp boot enabled we will not boot the DSP and firmware
up on system or rpm resume, just enable audio subsystem since audio IPs,
like HDA and SoundWire might be needed (codecs suspend/resume operation).
Only boot up the DSP during the first hw_params() call when the DSP is
really going to be needed.

In this way we can handle the audio related use cases:
normal audio use (rpm suspend/resume)
system suspend/resume without active audio
system suspend/resume with active audio
system suspend/resume without active audio, and audio start before the rpm
suspend timeout

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Intel: pci-lnl: Set on_demand_dsp_boot for LNL

LNL can be used with on-demand DSP booting, set the flag to enable it.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Intel: pci-ptl: Set on_demand_dsp_boot for PTL and WCL

PTL and WCL can be used with on-demand DSP booting, set the flag to
enable it.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v1:
- no need to check the on demand flag in hw_parms(), we have state check for that
- sof-probe client now works with on-demand DSP boot

- IPC message injector is not yet, neither compressed

ujfalusi commented:
Changes since v2:
- do a put module if the dsp boot returns with error in sof_client_core_module_get()

ujfalusi committed:
ASoC: SOF: ipc4-loader: Remove redundant rpm resume_and_get from load_library

The initial library loading is happening during topology loading, which is
already protected with pm_runtime_resume_and_get() via pcm.c

The redundant rpm code can be dropped from sof_ipc4_load_library()

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: control: skip rpm calls in ext_volatile_get if not implemented

Test earlier for the existence of ext_volatile_get callback and if it is
missing, skip the rpm calls to avoid needles DSP power on.

No change in functionality, we just skip the DSP power on in the unlikely
case when the ext_volatile _get is not supported and yet the topology adds
such control.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: sof-client: Add support for on-demand DSP boot

With the introduction of on-demand DSP boot the rpm status not necessary
tells that the DSP firmware is booted up.

Introduce the sof_client_boot_dsp() which can be used to make sure that
the DSP is booted and it can handle IPCs.

Update the client drivers to use the new function where it is expected that
the DSP is booted up.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v3:
- Cover corner case users to handle on-demand DSP boot mode: compressed, debug, ipc3-dtrace, client drivers.
- Ready for review.

bardliao commented on sound/soc/sof/pcm.c:
```diff
@@ -137,6 +137,16 @@ static int sof_pcm_hw_params(struct snd_soc_component *component,
+		 * Make sure that the DSP is booted up, which might not be the
+		 * case if the on-demand DSP boot is used
+		 */
+		ret = snd_sof_dsp_power_up(sdev, true);
```
I am thinking would it make more sense to call `snd_sof_dsp_power_up` in sof_pcm_prepare()?

ujfalusi replied:
I'm not 100% sure if we are not sending IPCs already in hw_params.
The prepare will call hw_params in case of system resume, probably it is OK to move the call to prepare, but I need to debug that to be sure.

bardliao replied:
Good point. sof_pcm_hw_params() uses IPC ops. We should not assume that it doesn't send IPCs in the ops.

kv2019i commented on sound/soc/sof/pm.c:
```diff
@@ -70,67 +70,41 @@ static void sof_cache_debugfs(struct snd_sof_dev *sdev)
-		return 0;
+			sof_set_fw_state(sdev, SOF_FW_BOOT_PREPARE);
+			snd_sof_dsp_pre_fw_run(sdev);
+			snd_sof_dsp_post_fw_run(sdev);
```
This pre/post pair looks a bit funky (given FW was not run). I guess you had some pressing reason to include here? A comment might help.

ujfalusi replied:
Yes, the pre and post fw run is used to do configuration which is needed to get the IPs powered, interrupts enabled, accesses configured.
I think I need them on pre LNL platforms, but I'm not sure anymore.

They do IP configuration around the booting up the DSP, I will add a comment.

kv2019i commented on sound/soc/sof/sof-client-ipc-flood-test.c:
```diff
@@ -219,9 +219,10 @@ static ssize_t sof_ipc_flood_dfs_write(struct file *file, const char __user *buf
 		goto out;
```
bad language in commit, "with introduction of on-demand DSP boot, the rpm status does not necessarily tell that the DSP firmware has booted up"

ujfalusi replied:
Sure, commit message is the hardest of all things!

kv2019i commented on sound/soc/sof/intel/mtl.c:
```diff
@@ -323,6 +323,9 @@ static int mtl_dsp_post_fw_run(struct snd_sof_dev *sdev)
 		}
```
The commit is a bit hard to follow " must be enabled since  ... they are only enabled" ?

ujfalusi replied:
right, what it is trying to say is that the interrupts on MTL are enabled as in step 4 of mtl_dsp_cl_init(), which is part of the firmware booting, but we need interrupts to be enabled for SDW/HDA to work, and since we don't boot the firmware, we need to do it on post fw run time.

lgirdwood replied:
can we inline comment this need to enable IRQs so its easier to follow.

ujfalusi replied:
I will drop the MTL support for now.

kv2019i commented on sound/soc/sof/intel/pci-mtl.c:
```diff
@@ -40,6 +40,7 @@ static const struct sof_dev_desc mtl_desc = {
 	.ipc_supported_mask	= BIT(SOF_IPC_TYPE_4),
 	.ipc_default		= SOF_IPC_TYPE_4,
 	.dspless_mode_supported	= true,		/* Only supported for HDaudio */
+	.on_demand_dsp_boot	= true,
```
Should we start with a incremental deployment before we expand to all platforms?

ujfalusi replied:
Yes, I was thinking of enabling this for LNL+ (ACE2+), which supports DSPless mode on all interfaces, unlike pre-LNL, where only HDA can be used in this way.

ujfalusi replied:
the MTL will be dropped from this for now.

kv2019i reviewed (commented):
Code looks good, lot of helpful explanations in commit messages. There are a few sentences in commits had hard language to follow. No blockers on my side, but given the magnitude of change, I wonder if we should do a partial deployment of this change (made the same comment inline)?

dbaluta commented:
@lgirdwood out of curiosity I wonder why this booting the firmware at system boot was needed. Beside a some sort of self-check at boot.

ujfalusi commented:
> @lgirdwood out of curiosity I wonder why this booting the firmware at system boot was needed. Beside a some sort of self-check at boot.

for self check, to be sure that the fw has the right key, to grab information about the booted firmware configuration and we are also placing it to IMR for faster bootup on consequent power ons.

lgirdwood commented on sound/soc/sof/sof-client-ipc-kernel-injector.c:
```diff
@@ -63,7 +63,9 @@ static ssize_t sof_kernel_msg_inject_dfs_write(struct file *file, const char __u
 	}
 
-	sof_client_ipc_rx_message(cdev, hdr, priv->kernel_buffer);
+	ret = sof_client_boot_dsp(cdev);
+	if (!ret)
+		sof_client_ipc_rx_message(cdev, hdr, priv->kernel_buffer);
```
can we embed the ```sof_client_boot_dsp()``` at the lowest level in the core IPC tx logic. This way we do it once.

ujfalusi replied:
I have thought about that, but we would still need a standalone API and decided against bundling it with IPC message sending.

lgirdwood replied:
We will always need that stand alone API, but we need to minimize where we call it. The natural place is in IPC since the host always initiates DSP use cases and it always does this with an IPC.

ujfalusi replied:
Unless the DSP must be booted up prior to few things. We want the DSP to be booted prior to DMA allocation as we don't know if we need DMA to reload the firmware or not, so tying with IPC is not going to work in most cases.
I rather not complicate the code with that as we will still need to keep most of the I need to have the DSP booted at this point calls anyways.

lgirdwood reviewed (commented):
Re BRA - this should only be used for active codec use cases. i.e. we are streaming audio.

@shumingfan @charleskeepax do we always need to load codec FW for jack detect use cases ? 

We should only need to power up codec and load codec FW when we have real streaming use cases otherwise at system resume we usually power up codec, load codec FW then after a few seconds enter RTD3 and power audio codec and DSP off again.  i.e. a wasted effort transition if audio is inactive.

lyakh commented on sound/soc/sof/pm.c:
```diff
@@ -154,16 +127,16 @@ static int sof_resume(struct device *dev, bool runtime_resume)
-		dev_warn(sdev->dev,
-			 "warning: failed to init trace after resume %d\n",
-			 ret);
+		dev_warn(sdev->dev, "%s: failed to resume trace%d\n",
```
a space is missing

ujfalusi replied:
Oh, will fix.

lyakh commented on sound/soc/sof/pm.c:
```diff
@@ -154,16 +127,16 @@ static int sof_resume(struct device *dev, bool runtime_resume)
 		ret = tplg_ops->set_up_all_pipelines(sdev, false);
 		if (ret < 0) {
-			dev_err(sdev->dev, "Failed to restore pipeline after resume %d\n", ret);
+			dev_err(sdev->dev, "%s: failed to restore pipeline%d\n",
```
ditto

lyakh commented on sound/soc/sof/compress.c:
```diff
@@ -195,6 +195,14 @@ static int sof_compr_set_params(struct snd_soc_component *component,
+	 * Make sure that the DSP is booted up, which might not be the
+	 * case if the on-demand DSP boot is used
+	 */
+	ret = snd_sof_dsp_power_up(sdev, true);
```
Do I understand it correctly, that currently powering up the DSP is a part of the (runtime) PM flow? I.e. is performed on some `.resume()` paths? And this commit makes additional locations like this one, which previously relied on the PM to bring up the DSP, do it explicitly now? Wouldn't it be possible and a better approach to keep the call as a part of the PM flow but just add a check in `.resume()` whether the DSP has to be booted?

ujfalusi replied:
@lyakh, currently we are powering up the audio subsystem _and_ booting up the DSP firmware, they are hard-linked.

With this PR we can detach the DSP firmware boot from the audio subsystem power up.

I have already reworked this API, I will have `snd_sof_boot_dsp_firmware(sdev);` only as a function, in pm.c it is:
```c
	if (sdev->pdata->desc->on_demand_dsp_boot) {
		/* Only change the fw_state to PREPARE but skip booting */
		sof_set_fw_state(sdev, SOF_FW_BOOT_PREPARE);
		return 0;
	}

	return snd_sof_boot_dsp_firmware(sdev);
```
for the resume.

ujfalusi committed:
ASoC: SOF: Intel: hda-sdw-bpt: Add support for on-demand DSP boot

If on-demand DSP boot is used we need to make sure that the DSP is booted
up - which might not be the case - since we need ChainDMA in normal, non
DSPless mode for the BRA to work.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

kv2019i reviewed (dismissed):
Looks great!

kv2019i reviewed (approved):
Looks good.

ranj063 commented:
@ujfalusi this PR is OK for now but I have a suggestion to make this a bit more elegant. How about we create a new auxiliary device for the DSP and use the runtime PM get/put functions to do the DSP boot or transition to D3 in those? That way instead of forcing the DSP to boot up, we can call the get/put to make sure that the DSP is booted when we need it in the params calls

lgirdwood commented:
> @ujfalusi this PR is OK for now but I have a suggestion to make this a bit more elegant. How about we create a new auxiliary device for the DSP and use the runtime PM get/put functions to do the DSP boot or transition to D3 in those? That way instead of forcing the DSP to boot up, we can call the get/put to make sure that the DSP is booted when we need it in the params calls

We do have time here, so if aligned and this request can be done easily then pls proceed.

ujfalusi commented:
> @ujfalusi this PR is OK for now but I have a suggestion to make this a bit more elegant. How about we create a new auxiliary device for the DSP and use the runtime PM get/put functions to do the DSP boot or transition to D3 in those? That way instead of forcing the DSP to boot up, we can call the get/put to make sure that the DSP is booted when we need it in the params calls

The problem is that it will look even more hackish as we have an aux device which does nothing, but to traverse back to use sdev and the audio device to boot up the DSP and then to turn it off.

What would be the clean solution is to have the audio device as it is now, create the aux devices for DSP and the interfaces, give them the resources t hat they need in order to be operational and use these devices instead of the audio device.
So, if the DSP is requested to be on then we would boot it up, if sdw is needed, we enable it, etc.

This also means  that the topology will not be loaded using the main sdev, but the auxdev of the DSP, this would traverse down to machine drivers, sdw stack, and other places as the device is no longer the audio device, but an auxdevice.

I'm not sure if hacking an auxdevice to hook back to the main device is cleaner than an API which does this.
It would be useful of the auxdev is used for the component and rely on ASoC to keep what's needed to be powered, but that would mean that we boot the DSP on resume likely.

While I think it would be great to do this, I honestly think that it will take a better part of the rest of the year to get it somehow right and then we have not even considered the vendors using SOF...

lgirdwood commented:
> While I think it would be great to do this, I honestly think that it will take a better part of the rest of the year to get it somehow right and then we have not even considered the vendors using SOF...

ok, if this is big then lets keep backport effort low.

ujfalusi commented:
Let's merge this for now to give broader testing
